config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java

```diff
@@ -345,33 +362,132 @@ private List<Endpoint> readEndpoints(Element parent, Optional<String> instance,
                 if (weightSum == 0) illegal(msgPrefix + "sum of all weights must be positive, got " + weightSum);
             } else {
                 if (stringAttribute("region", endpointElement).isPresent()) illegal(msgPrefix + "invalid 'region' attribute");
+                Set<RegionName> regions = new LinkedHashSet<>();
                 for (var regionElement : XML.getChildren(endpointElement, "region")) {
                     String region = regionElement.getTextContent();
                     if (region == null || region.isBlank()) illegal(msgPrefix + "empty 'region' element");
-                    targets.add(new Endpoint.Target(RegionName.from(region),
-                                                    InstanceName.from(instance.get()),
-                                                    1));
+                    if ( ! regions.add(RegionName.from(region))) illegal(msgPrefix + "duplicate 'region' element: '" + region + "'");
+                }
+
+                if (zoneEndpointType.isPresent()) {
+                    if (regions.isEmpty()) regions.add(null);
+                    ZoneEndpoint endpoint;
+                    Optional<String> enabled = XML.attribute("enabled", endpointElement);
+                    if ("zone".equals(zoneEndpointType.get())) {
+                        switch (enabled.orElse("true")) {
+                            case "true" -> endpoint = new ZoneEndpoint(true, false, List.of());
+                            case "false" -> endpoint = new ZoneEndpoint(false, false, List.of());
+                            default -> throw new IllegalArgumentException("Zone endpoint for container-id '" + containerId + "': " +
+                                                                          "invalid 'enabled' value; must be 'true' or 'false'");
+                        }
+                    }
+                    else {
+                        if (enabled.isPresent()) illegal("Private endpoint for container-id '" + containerId + "': " +
+                                                         "only endpoints of type 'zone' can specify 'enabled'");
+                        List<AllowedUrn> allowedUrns = new ArrayList<>();
+                        for (var allow : XML.getChildren(endpointElement, "allow")) {
```

Comment:
Attempting to set `<allow>` on a non-private endpoints should throw.

Reply:
Fixed, see above.